### PR TITLE
fetch-api-wrapper

### PR DESCRIPTION
**Pull request by Rajesh-Sundaram**
# Description
add fetch-api-wrapper to js-sdk to make http request at host-application.

# Checklist
- [X] I wrote/updated tests for new/changed code.
- [X] I ran `yarn sdk buildSdk` and `yarn bridge buildBridge` without issues.
- [X] I ran `yarn sample prettify` and `yarn sample build` without issues.

### CODE REVIEW

**Review by minh-rakuten — COMMENTED** [review]
It's so good but there is the issue of data type in cross platform.
I believe android & ios do not support `RequestInit` and `Response` of typescript so we need to define the data model in bridge.

The format of request and response you can take a look at `Mini App Web Requests to External APIs` internal document page. Please feel free to ping me or Cory if you can't find it.

File: js-miniapp-bridge/src/types/fetch.ts
Comment (corycaywood) [inline js-miniapp-bridge/src/types/fetch.ts:2]: I agree with @minh-rakuten about this - We should explicitly specify our own type rather than relying on part of the browser's/typescript's implementation. We are using the browser's `fetch` as inspiration, but there will end up being some differences.

File: js-miniapp-bridge/src/types/fetch.ts
Comment (corycaywood) [inline js-miniapp-bridge/src/types/fetch.ts:3]: We're starting out with support only for `GET` requests, so there's no need for `body` currently, right?

File: js-miniapp-bridge/src/types/fetch.ts
Comment (corycaywood) [inline js-miniapp-bridge/src/types/fetch.ts:1]: How about the name `FetchRequest`? I don't think we need to add the `MiniApp` prefix to our types because the user will be able rename them when importing, if needed: i.e. `import {WebRequest as MiniAppRequest} from 'js-miniapp-sdk'`

File: js-miniapp-sdk/src/miniapp.ts
Comment (corycaywood) [inline js-miniapp-sdk/src/miniapp.ts:47]: I think we should explicitly define the `Response` type too. Maybe it can be named `FetchResponse`?

Also, shouldn't `options` be optional? I wonder if we should use the same style as the browser's `fetch` where it can be called as any of the following:
- `fetch("https://example.com")`
- `fetch("https://example.com", {headers: headers})`
- `fetch({url: "https://example.com", headers: headers})`

What do you think?

**Review by corycaywood — COMMENTED** [review]
It's looking good 👍  I've left a few comments.

File: js-miniapp-sdk/src/miniapp.ts
Comment (Abhijit-majumder) [reply js-miniapp-sdk/src/miniapp.ts:47]: I agree with @corycaywood here that we should explicitly define the response interface.

File: js-miniapp-bridge/src/types/fetch.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-bridge/src/types/fetch.ts]: If we don't have any security implications on providing POST method support, can we have the POST method implementation on the native-side with a simple string (raw-data, JSON) as a request body for the initial-phase?

Comment (corycaywood) [reply js-miniapp-bridge/src/types/fetch.ts]: Hmm, I'm not sure - there aren't really any security concerns, but this will add more complexity on the native side, which we wanted to avoid for this initial implementation.

On the JS side, it's not very difficult to add support for POST requests, so maybe we can leave it out for now and then it will still be easy to add in JS when we decide to support it on the native side.

File: js-miniapp-bridge/src/common-bridge.ts
Comment (corycaywood) [inline js-miniapp-bridge/src/common-bridge.ts:288]: I think `error` should pass a string and not a `Response`. This is because the Promise will only be rejected if there was some issue with sending the request or receiving a response (i.e. the user set invalid params, tried to connect to a URL which isn't in the "Allow List", device is not connected to network, etc.) So in that case, there will be no `Response` object to return to the user.

If the server returns an error response (i.e. 500), then the Promise will be resolved with the error response - this is why the `Response` model has an `ok` property to check if the response is successful.

**Review by corycaywood — COMMENTED** [review]
Looking good! 👍  I left some comments below.

File: js-miniapp-bridge/src/types/fetch.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-bridge/src/types/fetch.ts:27]: 👍

File: js-miniapp-sdk/src/miniapp.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-sdk/src/miniapp.ts:195]: FetchRequestInit doesn't have `url` property. init option can be used to override the headers & method.
if we ignore the second arg, can we throw an error/warning message when the user passes both object args?

File: js-miniapp-sdk/src/miniapp.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-sdk/src/miniapp.ts:187]: in progress 👍

File: js-miniapp-bridge/src/common-bridge.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-bridge/src/common-bridge.ts:288]: 👍  makes sense. Thanks

File: js-miniapp-bridge/src/types/fetch.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-bridge/src/types/fetch.ts]: would it be okay if we keep this implementations of bridge-types in js-SDK workspace?

File: js-miniapp-sdk/src/miniapp.ts
Comment (corycaywood) [reply js-miniapp-sdk/src/miniapp.ts]: Hmm, maybe we should just simplify things and only support `fetch(input: string, opts?: FetchRequestInit)`? 

Do you think there are any benefits to supporting `fetch(input: FetchRequest | string, opts?: FetchRequestInit)`? The only thing I can think of is similarity to the browser's `fetch` function, but our `fetch` will already be different anyways.

File: js-miniapp-bridge/src/types/fetch.ts
Comment (corycaywood) [reply js-miniapp-bridge/src/types/fetch.ts]: I don't think that the `sdk` workspace is the right place for this. The `js-miniapp-sdk` npm module is mainly intended as a wrapper for the bridge functionality, so we try not to put much in there beyond what we have to.

Also, keeping this in the `bridge` workspace couples it with the native side because the bridge JS gets bundled with the native SDK. So if we make changes to the native side then we can simultaneously makes changes to the bridge JS. This is important in case we need to make changes in the bridge which would otherwise be breaking for mini apps that included an old version of `js-miniapp-sdk`.

File: js-miniapp-sdk/src/miniapp.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-sdk/src/miniapp.ts]: I absolutely agree with this simplification 👍

File: js-miniapp-bridge/src/types/fetch.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-bridge/src/types/fetch.ts]: 👍 Got it. Thanks for clarifying

File: js-miniapp-bridge/src/fetch/response.ts
Comment (minh-rakuten) [inline js-miniapp-bridge/src/fetch/response.ts:10]: Why do we have to declare again when `Response` and `ResponeBody` have them already?

**Review by minh-rakuten — COMMENTED** [review]
I just only have questions due to my lack of knowledge.

**Review by corycaywood — COMMENTED** [review]
Looks good - just a few issues with the public interface and generated docs.

File: js-miniapp-bridge/src/fetch/response.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-bridge/src/fetch/response.ts:10]: since DecodedFetchResponse extends the Response type, it's required to implement all of its parent properties in its class.

File: js-miniapp-sdk/src/miniapp.ts
Comment (Rajesh-Sundaram) [reply js-miniapp-sdk/src/miniapp.ts:185]: 👍  async/await not required as this bridge returns promise
updated with below
`fetch(url: string, opts?: FetchRequestInit): Promise<FetchResponse> {
    return this.bridge
      .fetch(new InternalFetchRequest(url, opts))
      .then(res => new DecodedFetchResponse(res));
  }`

**Review by minh-rakuten — APPROVED** [review]
👍 
There could be the update for this when native sdk implements this feature but we will figure it out later.

**Review by corycaywood — COMMENTED** [review]
Looks good 👍 

But I think we shouldn't merge this yet because I'm not sure when we want to release this. Or maybe you can create a new feat/fetch branch for merging this PR there?

**Review by Abhijit-majumder — APPROVED** [review]
Looks Good 👍

**Comment by Rajesh-Sundaram** [comment]
> Looks good 👍
> 
> But I think we shouldn't merge this yet because I'm not sure when we want to release this. Or maybe you can create a new feat/fetch branch for merging this PR there?

@corycaywood 👍 Done. Can I merge?

**Review by corycaywood — APPROVED** [review]
👍

**Pull request merged**
